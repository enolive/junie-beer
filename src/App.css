body {
  font-family:
    -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
    'Helvetica Neue', sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  background: linear-gradient(135deg, #f5f7fa 0%, #c3cfe2 100%);
  min-height: 100vh;
}

.App {
  max-width: 1200px;
  margin: 0 auto;
  padding: 20px;
}

.App-header {
  text-align: center;
  margin-bottom: 40px;
  padding: 30px;
  background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
  border-radius: 15px;
  color: white;
  box-shadow: 0 8px 32px rgba(0, 0, 0, 0.1);
}

.App-header h1 {
  font-size: 2.5rem;
  margin-bottom: 10px;
  text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
}

.App-header p {
  font-size: 1.2rem;
  opacity: 0.9;
}

.App-main {
  display: grid;
  grid-template-columns: 1fr 2fr;
  gap: 30px;
  align-items: start;
}

@media (max-width: 768px) {
  .App-main {
    grid-template-columns: 1fr;
  }
}

.App-footer {
  text-align: center;
  margin-top: 40px;
  padding: 20px;
  background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
  border-radius: 15px;
  color: white;
  box-shadow: 0 8px 32px rgba(0, 0, 0, 0.1);
  opacity: 0.9;
}

.App-footer p {
  margin-bottom: 10px;
  font-size: 1rem;
}

.App-footer a {
  color: white;
  text-decoration: none;
  font-weight: 600;
  border-bottom: 2px solid rgba(255, 255, 255, 0.5);
  padding-bottom: 2px;
  transition: border-color 0.3s ease;
}

.App-footer a:hover {
  border-color: white;
}

.add-beer-section {
  background: white;
  padding: 30px;
  border-radius: 15px;
  box-shadow: 0 8px 32px rgba(0, 0, 0, 0.1);
  height: fit-content;
}

.add-beer-section h2 {
  color: #333;
  margin-bottom: 25px;
  font-size: 1.5rem;
  border-bottom: 3px solid #667eea;
  padding-bottom: 10px;
}

.rating-text {
  display: flex;
  gap: 6px;
}

.submit-btn {
  background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
  color: white;
  border: none;
  padding: 15px 30px;
  border-radius: 8px;
  font-size: 1.1rem;
  font-weight: 600;
  cursor: pointer;
  transition:
    transform 0.2s ease,
    box-shadow 0.2s ease;
}

.submit-btn:hover {
  transform: translateY(-2px);
  box-shadow: 0 8px 25px rgba(102, 126, 234, 0.3);
}

.submit-btn:active {
  transform: translateY(0);
}

.beer-list-section {
  background: white;
  padding: 30px;
  border-radius: 15px;
  box-shadow: 0 8px 32px rgba(0, 0, 0, 0.1);
}

.beer-list-section h2 {
  color: #333;
  margin-bottom: 25px;
  font-size: 1.5rem;
  border-bottom: 3px solid #667eea;
  padding-bottom: 10px;
}

.no-beers {
  text-align: center;
  color: #666;
  font-style: italic;
  padding: 40px;
  background: #f8f9fa;
  border-radius: 10px;
  border: 2px dashed #ddd;
}

.beer-grid {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  gap: 20px;
  list-style: none;
  padding: 0;
  margin: 0;
}

.beer-card dl {
  margin: 0;
  display: grid;
  grid-template-columns: auto 1fr;
  gap: 8px 16px;
  align-items: baseline;
}

.beer-card dt {
  font-weight: bold;
  color: #555;
  grid-column: 1;
}

.beer-card dd {
  margin: 0;
  grid-column: 2;
}

.beer-card {
  background: linear-gradient(135deg, #e0eafc 0%, #cfdef3 100%);
  box-shadow: 0 4px 15px rgba(0, 0, 0, 0.1);
  transition:
    transform 0.2s ease,
    box-shadow 0.2s ease;
}

.beer-card:hover {
  transform: translateY(-5px);
  box-shadow: 0 8px 25px rgba(0, 0, 0, 0.15);
}

/* Responsive design */
@media (max-width: 768px) {
  .App {
    padding: 15px;
  }

  .App-header {
    padding: 20px;
  }

  .App-header h1 {
    font-size: 2rem;
  }

  .add-beer-section,
  .beer-list-section {
    padding: 20px;
  }

  .beer-grid {
    grid-template-columns: 1fr;
  }
}

@media (max-width: 480px) {
  .App-header h1 {
    font-size: 1.8rem;
  }
}
