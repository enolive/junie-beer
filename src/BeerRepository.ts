// BeerRepository.ts

import { Beer } from './Domain.ts'

const STORAGE_KEY = 'junies-beer-tracker'

export class BeerRepository {
  loadBeers(): Beer[] {
    const savedBeers = localStorage.getItem(STORAGE_KEY)
    return savedBeers ? JSON.parse(savedBeers) : []
  }

  saveBeers(beers: Beer[]): void {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(beers))
  }
}

export const beerRepository = new BeerRepository()
