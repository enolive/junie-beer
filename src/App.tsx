import React, { useEffect, useState } from 'react'
import BeerForm, { BeerFormData } from './BeerForm'
import BeerList from './BeerList'
import { beerRepository } from './BeerRepository'
import { Beer } from './Domain.ts'

interface AppProps {
  dateFn?: () => Date
}

const App: React.FC<AppProps> = ({ dateFn }) => {
  const [beers, setBeers] = useState<Beer[]>(() => beerRepository.loadBeers())
  const getDate = dateFn || (() => new Date())

  useEffect(() => {
    beerRepository.saveBeers(beers)
  }, [beers])

  const handleBeerSubmit = (beerData: BeerFormData) => {
    const isoDate = getDate().toISOString().split('T')[0]
    const beer: Beer = {
      ...beerData,
      id: 0,
      dateAdded: isoDate,
    }
    setBeers(prev => [...prev, { ...beer, id: prev.length + 1 }])
  }

  const handleDeleteBeer = (id: number) => {
    setBeers(prev => prev.filter(beer => beer.id !== id))
  }

  return (
    <div className="max-w-7xl mx-auto px-5">
      <header className="text-center mb-10 p-8 gradient-background rounded-2xl shadow-lg text-white mt-5">
        <h1 className="text-4xl font-bold mb-2 drop-shadow-md">üç∫ Junie's Beer Tracker</h1>
        <p role="complementary" className="text-lg opacity-90">
          Keep track of your favorite beers!
        </p>
      </header>

      <main className="grid grid-cols-1 md:grid-cols-3 gap-8">
        <section
          role="region"
          aria-labelledby="beer-form-heading"
          className="md:col-span-1 bg-white p-8 rounded-2xl shadow-lg"
        >
          <h2 id="beer-form-heading" className="section-header">
            Add New Beer
          </h2>
          <BeerForm onSubmit={handleBeerSubmit} />
        </section>

        <section
          role="region"
          aria-labelledby="beer-collection-heading"
          className="md:col-span-2 bg-white p-8 rounded-2xl shadow-lg"
        >
          <h2 id="beer-collection-heading" className="section-header">
            Your Beer Collection ({beers.length})
          </h2>
          <BeerList beers={beers} onDeleteBeer={handleDeleteBeer} labelId="beer-collection-heading" />
        </section>
      </main>

      <footer className="text-center mt-10 mb-5 p-6 gradient-background rounded-2xl shadow-lg text-white">
        <p className="mb-2 opacity-90">
          Created with the <b>help</b> of the almighty AI
        </p>
        <a
          href="https://www.jetbrains.com/junie"
          target="_blank"
          rel="noopener noreferrer"
          className="text-white font-semibold border-b-2 border-white/50 hover:border-white transition-colors"
        >
          Visit Junie AI's website
        </a>
      </footer>
    </div>
  )
}

export default App
