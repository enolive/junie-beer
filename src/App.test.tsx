import { render, screen } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import App from './App'
import { setupFormActions } from './test/setup-form-actions'
import { setupListActions } from './test/setup-list-actions'
import { beerRepository } from './BeerRepository.ts'

vi.mock('./BeerRepository')

describe('App', () => {
  beforeEach(() => {
    mockedLoadBeers.mockClear()
    mockedLoadBeers.mockReturnValue([])
  })

  it('renders the main header and sections', () => {
    renderComponent()

    expect(screen.getByRole('banner')).toHaveTextContent("🍺 Junie's Beer Tracker")
    expect(screen.getByRole('complementary')).toHaveTextContent('Keep track of your favorite beers!')
    expect(screen.getByRole('region', { name: 'Add New Beer' })).toBeInTheDocument()
    expect(screen.getByRole('region', { name: 'Your Beer Collection (0)' })).toBeInTheDocument()
  })

  it('shows empty state message when no beers are added', () => {
    renderComponent()

    expect(screen.getByText('No beers added yet. Add your first beer above!')).toBeInTheDocument()
  })

  it('adds a beer with required fields only', async () => {
    renderComponent()
    const actions = setupFormActions()

    await actions.fillAllFields({ name: 'Test IPA', brewery: 'Test Brewery' })
    await actions.submitForm()

    expect(screen.getByText('Your Beer Collection (1)')).toBeInTheDocument()
    expect(screen.getByText('Test IPA')).toBeInTheDocument()
    expect(screen.getByText('Test Brewery')).toBeInTheDocument()
    expect(screen.queryByText('No beers added yet')).not.toBeInTheDocument()
  })

  it('adds a beer with all fields filled', async () => {
    renderComponent()
    const actions = setupFormActions()

    await actions.fillAllFields({
      name: 'Hoppy IPA',
      brewery: 'Craft Brewery',
      style: 'IPA',
      notes: 'Very hoppy and delicious',
      rating: 4,
    })
    await actions.submitForm()

    expect(screen.getByText('Your Beer Collection (1)')).toBeInTheDocument()
    expect(screen.getByText('Hoppy IPA')).toBeInTheDocument()
    expect(screen.getByText('Craft Brewery')).toBeInTheDocument()
    expect(screen.getByText('IPA')).toBeInTheDocument()
    expect(screen.getByText('Very hoppy and delicious')).toBeInTheDocument()
    expect(screen.getByText('⭐⭐⭐⭐ (4/5)')).toBeInTheDocument()
  })

  it('deletes a beer when delete button is clicked', async () => {
    renderComponent()
    const formActions = setupFormActions()
    const listActions = setupListActions()

    await formActions.fillAllFields({ name: 'Beer to Delete', brewery: 'Delete Brewery' })
    await formActions.submitForm()

    expect(screen.getByText('Your Beer Collection (1)')).toBeInTheDocument()
    expect(screen.getByText('Beer to Delete')).toBeInTheDocument()

    await listActions.deleteBeer('Beer to Delete')

    expect(screen.getByText('Your Beer Collection (0)')).toBeInTheDocument()
    expect(screen.queryByText('Beer to Delete')).not.toBeInTheDocument()
    expect(screen.getByText('No beers added yet. Add your first beer above!')).toBeInTheDocument()
  })

  it('adds multiple beers and displays correct count', async () => {
    renderComponent()
    const actions = setupFormActions()

    await actions.fillAllFields({ name: 'First Beer', brewery: 'First Brewery' })
    await actions.submitForm()
    await actions.fillAllFields({ name: 'Second Beer', brewery: 'Second Brewery' })
    await actions.submitForm()
    await actions.fillAllFields({ name: 'Third Beer', brewery: 'Third Brewery' })
    await actions.submitForm()

    expect(screen.getByText('Your Beer Collection (3)')).toBeInTheDocument()
    expect(screen.getByText('First Beer')).toBeInTheDocument()
    expect(screen.getByText('Second Beer')).toBeInTheDocument()
    expect(screen.getByText('Third Beer')).toBeInTheDocument()
  })

  it('loads beers from localStorage on initialization', () => {
    const mockBeers = [
      {
        id: 1,
        name: 'Saved Beer',
        brewery: 'Saved Brewery',
        style: 'IPA',
        rating: 4,
        notes: 'Loaded from storage',
        dateAdded: '1/1/2024',
      },
    ]
    mockedLoadBeers.mockReturnValueOnce(mockBeers)

    renderComponent()

    expect(screen.getByText('Your Beer Collection (1)')).toBeInTheDocument()
    expect(screen.getByText('Saved Beer')).toBeInTheDocument()
    expect(screen.getByText('Saved Brewery')).toBeInTheDocument()
  })

  it('saves beers to localStorage when a beer is added', async () => {
    const mockBeers = [
      {
        id: 1,
        name: 'Saved Beer',
        brewery: 'Saved Brewery',
        style: 'IPA',
        rating: 4,
        notes: 'Loaded from storage',
        dateAdded: '1/1/2024',
      },
    ]
    mockedLoadBeers.mockReturnValueOnce(mockBeers)
    renderComponent()
    const actions = setupFormActions()

    await actions.fillAllFields({
      name: 'New Beer',
      brewery: 'New Brewery',
      rating: 4,
      notes: 'Great beer!',
      style: 'IPA',
    })
    await actions.submitForm()

    expect(mockedSaveBeers).toHaveBeenCalledWith([
      ...mockBeers,
      {
        id: 2,
        name: 'New Beer',
        brewery: 'New Brewery',
        style: 'IPA',
        rating: 4,
        notes: 'Great beer!',
        dateAdded: '11.7.2025',
      },
    ])
  })

  it('updates localStorage when a beer is deleted', async () => {
    const mockBeers = [
      {
        id: 123,
        name: 'Beer to Delete',
        brewery: 'Delete Brewery',
        style: '',
        rating: 0,
        notes: '',
        dateAdded: '1/1/2024',
      },
    ]
    mockedLoadBeers.mockReturnValueOnce(mockBeers)
    renderComponent()
    const listActions = setupListActions()

    await listActions.deleteBeer('Beer to Delete')

    expect(mockedSaveBeers).toHaveBeenCalled()
    const lastCall = mockedSaveBeers.mock.calls[mockedSaveBeers.mock.calls.length - 1]
    const savedBeers = lastCall[0]
    expect(savedBeers).toHaveLength(0)
  })
})

function renderComponent() {
  render(<App dateFn={() => new Date('2025-07-11T00:00Z')} />)
}

const mockedLoadBeers = vi.mocked(beerRepository.loadBeers)
const mockedSaveBeers = vi.mocked(beerRepository.saveBeers)
