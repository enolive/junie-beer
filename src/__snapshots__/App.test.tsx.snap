// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders the footer with correct text and link 1`] = `
<footer
  class="App-footer"
>
  <p>
    Created with the 
    <b>
      help
    </b>
     of the almighty AI
  </p>
  <a
    href="https://www.jetbrains.com/junie"
    rel="noopener noreferrer"
    target="_blank"
  >
    Visit Junie AI's website
  </a>
</footer>
`;
